pyccl/cells.py

```diff
@@ -22,35 +33,58 @@ def angular_cl(cosmo, tracer1, tracer2, ell, *,
-            approximation will be used.
-        limber_integration_method (:obj:`str`): integration method to be used
-            for the Limber integrals. Possibilities: ``'qag_quad'`` (GSL's
-            `qag` method backed up by `quad` when it fails) and ``'spline'``
-            (the integrand is splined and then integrated numerically).
+            the non-linear power spectra stored in `cosmo` (e.g.
+            `'delta_matter:delta_matter'`).
+        l_limber (int, float or 'auto') : Angular wavenumber beyond which 
+            Limber's approximation will be used. Defaults to -1. If 'auto', 
+            then the non-limber integrator will be used to compute the right 
+            transition point given the value of limber_max_error.
+        limber_max_error (float) : Maximum fractional error for Limber integration.
+        limber_integration_method (string) : integration method to be used
+            for the Limber integrals. Possibilities: 'qag_quad' (GSL's `qag`
+            method backed up by `quad` when it fails) and 'spline' (the
+            integrand is splined and then integrated numerically).
+        non_limber_integration_method (string) : integration method to be used
+            for the non-Limber integrals. Possibilities: 'FKEM','MATTER'.
+            See N5K (arXiv:2212.04291) paper for more information.
+        return_meta (bool): if `True`, also return a dictionary with various
+            metadata about the calculation, such as l_limber as calculated by the
+            non-limber integrator.
 
     Returns:
         :obj:`float` or `array`: Angular (cross-)power spectrum values, \
             :math:`C_\\ell`, for the pair of tracers, as a function of \
             :math:`\\ell`.
-    """ # noqa
-    if cosmo['Omega_k'] != 0:
+    """  # noqa
+    if cosmo["Omega_k"] != 0:
         warnings.warn(
             "CCL does not properly use the hyperspherical Bessel functions "
             "when computing angular power spectra in non-flat cosmologies!",
-            category=CCLWarning)
+            category=CCLWarning,
+        )
+
     if limber_integration_method not in integ_types:
         raise ValueError(
-            f"Unknown integration method {limber_integration_method}.")
+            "Limber integration method %s not supported"
+            % limber_integration_method
+        )
+    if non_limber_integration_method not in ["FKEM", "MATTER"]:
+        raise ValueError(
+            "Non-Limber integration method %s not supported"
+            % limber_integration_method
+        )
+    if type(l_limber) == str:
+        if l_limber != "auto":
+            raise ValueError("l_limber cannot be a string other than 'auto'")
+        auto_limber = True
+    else:
+        auto_limber = False
 
     # we need the distances for the integrals
     cosmo.compute_distances()
 
     # Access ccl_cosmology object
     cosmo_in = cosmo
-    cosmo = cosmo.cosmo
```

Comment:
This whole cosmo_in vs. cosmo makes things confusing (this is on me, probably). Can we just use `cosmo` for the python object and `cosmo.cosmo` for the C structure everywhere?